proof/bls_operations.saw

```diff
@@ -0,0 +1,1093 @@
+    simplify (basic_ss);
+    simplify (cryptol_ss());
+    simplify (addsimps core_rewrites empty_ss);
+    simplify (addsimps split_join_thms empty_ss);
+    simplify (addsimps [ deserialize_E2_uncompress_E2_OK_thm ] empty_ss);
+    simplify remove_higher_order_function_simpset;
+    goal_num_ite 4
+      (do {
+        simplify (addsimp just_thm empty_ss);
+        simplify (basic_ss);
+        simplify (cryptol_ss());
+        simplify (addsimps [ fold_normalize_affine_fp2_thm
+                           , hoist_pair_thm
+                           , normalize_point_O_E'_thm
+                           , normalize_uncompress_E2_OK_thm
+                           , hoist_normalize_affine_fp2_thm ] empty_ss);
+        simplify (basic_ss);
+        simplify (cryptol_ss());
+        (print_goal);
+        w4_unint_z3 [ "is_in_g2_impl"
+                    , "uncompress_E2_OK"
+                    , "is_point_affine_E'"
+                    , "pubkey_subgroup_check"
+                    , "is_point_O_E'"
+                    , "is_point_affine_E'"
+                    , "deserialize_E2"
+                    , "POINTonE2_affine_invariant"
+                    , "fp2_rep"
+                    , "uncompress_E2_imp"];
+      })
+      (do { 
+        (print_goal);
+        w4_unint_z3 [ "is_in_g2_impl"
+                    , "uncompress_E2_OK"
+                    , "is_point_affine_E'"
+                    , "pubkey_subgroup_check"
+                    , "is_point_O_E'"
+                    , "is_point_affine_E'"
+                    , "deserialize_E2"
+                    , "uncompress_E2_imp"];
+       });
+  };
+
+let demo_BasicSign_A_spec = do {
+  //let DST = {{ "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_" }};
+  global_alloc_init "demo_DST_A" {{ DST }};
+  let msg_len = 32;
+  out_ptr <- llvm_alloc (llvm_array 48 (llvm_int 8));
+  // SK must be aligned
+  SK_ptr <- llvm_alloc_readonly_aligned 8 pow256_type;
+  SK <- llvm_fresh_var "SK" pow256_type;
+  llvm_points_to SK_ptr (llvm_term SK);
+  (msg, msg_ptr) <- ptr_to_fresh_readonly "msg" (llvm_array msg_len (llvm_int 8));
+  // TODO: This precond was the only way I could think of to be able to talk
+  // about the relationship between SK and the result of blst_hash_to_g1.  Is
+  // there maybe a theorem that would work better?
+  llvm_precond {{ e1_order (affinify E (hash_to_curve_opt_impl (msg,DST))) >
+                  scalar_value`{255,32} SK + shift }};
+  // TODO: Do I need this precond VV ?
+  llvm_precond {{ SK != zero }};
+  llvm_execute_func [ out_ptr
+                    , SK_ptr
+                    , msg_ptr
+                    , llvm_term {{ `msg_len : [64] }} ];
+  // TODO: Endianness?  (Both for return, and for args to CoreSign)
+  llvm_points_to
+    out_ptr
+    (llvm_term {{ split`{each=8} (CoreSign (drop_join_reverse`{255, 32} SK,
+                                            msg)) }});
+};
+
+// TODO: Under what conditions is this true?  Should probably be done over IETF
```

Comment:
I think some version of this is proved using an assumption about the high-level spec in `hash_to_g2.saw`